src/dparse/ast.d

```diff
@@ -1909,6 +1925,18 @@ final class IndexExpression : ExpressionNode
     mixin OpEquals;
 }
 
+///
+final class InContractExpression : ASTNode
+{
+    override void accept(ASTVisitor visitor) const
+    {
+        mixin (visitIfNotNull!(assertArguments));
+    }
+    /** */ size_t inTokenLocation;
```

Comment:
I don't think this is required. IIRC you wanted this for DFMT but with the official grammar FunctionContract gives the information if it's a expression style contract (its InOutStatement is null) or a statement style contract (its InOutContractExpression) is null.

Reply:
`inTokenLocation` and `outTokenLocation` were already used for detecting classic contract statements; the problem with them was simply that they weren't filled for expression-based contracts, so DFMT didn't consider them as contracts.
But `inTokenLocation` and `outTokenLocation` are used to distinguish contracts from other usages of `in` and `out`, and not to distinguish expression-based ones from statement ones, so I think it's probably better to leave them there

Reply:
ok then. since my comment i just remembered too that DFMT is based on the tokens and only very partially the AST.